core/trino-main/src/main/java/io/trino/server/ui/LoginPreviewResource.java

```diff
@@ -0,0 +1,118 @@
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.HttpHeaders;
+import jakarta.ws.rs.core.NewCookie;
+import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.core.SecurityContext;
+
+import java.util.Optional;
+
+import static com.google.common.base.Strings.emptyToNull;
+import static io.trino.server.security.ResourceSecurity.AccessType.WEB_UI;
+import static io.trino.server.ui.FormWebUiAuthenticationFilter.UI_PREVIEW_AUTH_INFO;
+import static io.trino.server.ui.FormWebUiAuthenticationFilter.UI_PREVIEW_LOGIN;
+import static io.trino.server.ui.FormWebUiAuthenticationFilter.UI_PREVIEW_LOGOUT;
+import static io.trino.server.ui.FormWebUiAuthenticationFilter.getDeleteCookies;
+import static jakarta.ws.rs.core.MediaType.APPLICATION_JSON;
+import static java.util.Objects.requireNonNull;
+
+@Path("")
+public class LoginPreviewResource
+        implements AuthPreviewResource
+{
+    private final FormWebUiAuthenticationFilter formWebUiAuthenticationManager;
+
+    @Inject
+    public LoginPreviewResource(FormWebUiAuthenticationFilter formWebUiAuthenticationManager)
+    {
+        this.formWebUiAuthenticationManager = requireNonNull(formWebUiAuthenticationManager, "formWebUiAuthenticationManager is null");
+    }
+
+    @Override
+    @ResourceSecurity(WEB_UI)
+    @GET
+    @Path(UI_PREVIEW_AUTH_INFO)
+    @Produces(APPLICATION_JSON)
+    public AuthInfo getAuthInfo(ContainerRequestContext request, @Context SecurityContext securityContext)
+    {
+        boolean isPasswordAllowed = formWebUiAuthenticationManager.isPasswordAllowed(securityContext.isSecure());
+        Optional<String> username = formWebUiAuthenticationManager.getAuthenticatedUsername(request);
+        return new AuthInfo("form", isPasswordAllowed, username.isPresent(), username);
+    }
+
+    @ResourceSecurity(WEB_UI)
+    @POST
+    @Path(UI_PREVIEW_LOGIN)
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    public Response login(LoginForm loginForm, @Context SecurityContext securityContext)
+    {
+        final String username = emptyToNull(loginForm.username);
+        final String password = emptyToNull(loginForm.password);
+
+        if (!formWebUiAuthenticationManager.isAuthenticationEnabled(securityContext.isSecure())) {
+            throw new ForbiddenException();
+        }
+
+        Optional<NewCookie[]> authenticationCookie = formWebUiAuthenticationManager.checkLoginCredentials(username, password, securityContext.isSecure());
+        if (authenticationCookie.isEmpty()) {
+            throw new ForbiddenException();
+        }
+
+        return Response.ok("{}")
+                .cookie(authenticationCookie.get())
+                .build();
+    }
+
+    @ResourceSecurity(WEB_UI)
+    @GET
+    @Path(UI_PREVIEW_LOGOUT)
+    @Produces(APPLICATION_JSON)
+    public Response logout(@Context HttpHeaders httpHeaders, @Context SecurityContext securityContext)
+    {
+        return Response.ok("{}")
```

Comment:
noContent() + redirect?

Reply:
fixed but no redirect at the end. This endpoint is intended for APIs and the single-page frontend app so direct redirection isn’t required. The React app manages re-rendering the login form independently.